src/node.cpp

```diff
@@ -26,56 +26,68 @@
 
 namespace dpaste {
 
-std::uniform_int_distribution<dht::Value::Id> vid_dist;
-std::mt19937_64 rand_;
+Node::Node()
+{
+    dht::crypto::random_device rdev;
+    std::seed_seq seed {rdev(), rdev()};
+    rand_.seed(seed);
+}
+
+std::string Node::paste(dht::Blob&& blob, dht::DoneCallbackSimple&& cb) {
+    auto pin = codeDist_(rand_);
+    std::stringstream ss;
+    ss << std::hex << pin;
+    auto pin_str = ss.str();
+    std::transform(pin_str.begin(), pin_str.end(), pin_str.begin(), ::toupper);
 
-dht::InfoHash Node::paste(dht::Blob&& blob, dht::DoneCallbackSimple&& cb) {
-	auto h = dht::InfoHash::getRandom();
-	paste(h, std::forward<dht::Blob>(blob), std::forward<dht::DoneCallbackSimple>(cb));
-	return h;
+    paste(pin_str, std::forward<dht::Blob>(blob), std::forward<dht::DoneCallbackSimple>(cb));
+    return pin_str;
 }
 
-void Node::paste(dht::InfoHash hash, dht::Blob&& blob, dht::DoneCallbackSimple&& cb) {
-	auto v = std::make_shared<dht::Value>(std::forward<dht::Blob>(blob));
-	v->id = vid_dist(rand_);
-	v->type = dht::ValueType::USER_DATA.id;
+void Node::paste(const std::string& code, dht::Blob&& blob, dht::DoneCallbackSimple&& cb) {
+    auto v = std::make_shared<dht::Value>(std::forward<dht::Blob>(blob));
     v->user_type = DPASTE_USER_TYPE;
 
-	if (cb)
-		node.put(hash, v, cb);
-	else {
-		auto mtx = std::make_shared<std::mutex>();
-		auto cv = std::make_shared<std::condition_variable>();
-		std::unique_lock<std::mutex> lk(*mtx);
-		node.put(hash, v, [&hash,&mtx,&cv] (bool success) {
-			std::unique_lock<std::mutex> lk(*mtx);
-			if (not success)
-				/* TODO: use better logging methods */
-				std::cerr << CONNECTION_FAILURE_MSG << std::endl;
-			cv->notify_all();
-		});
-		cv->wait(lk);
-	}
+    auto hash = dht::InfoHash::get(code);
+
+    if (cb)
+        node.put(hash, v, cb);
+    else {
+        std::mutex mtx;
+        std::condition_variable cv;
+        std::unique_lock<std::mutex> lk(mtx);
+        bool done {false};
```

Comment:
As per the standard: "wait causes the current thread to block until the condition variable is notified or a spurious wakeup occurs". Spurious wakeup can occur even without any call to notify.